### PR TITLE
Feature/yu mei/cns 33 dynamodb for user info

### PR DESCRIPTION
**Pull request by amyhsu-o**
## Changes

- DynamoDB service for user info
- register and login logic in authService related to user info in DynamoDB

## Setup

get access key in aws console 
1. search for IAM
2. click 'Users' in IAM resources
3. click 'Clouddrop-DynamoDB'
4. click 'Create access key'
5. follow `.env.example` to make the `.env` , and put the access key in it

## Test

Start backend server

```bash
cd backend
npm install
npm start
``` 

Use postman

### Test 1:  /register
image

### Test 2: /login
image

You can go to [DynamoDB in aws console](https://ap-northeast-1.console.aws.amazon.com/dynamodbv2/home?region=ap-northeast-1#item-explorer?index=userEmail&maximize=true&operation=SCAN&table=userInfoRecord) to see the new data added in through the API. There are currently some demonstration data in it. Click 'Run' to make it refresh for your new data.

For user info, the structure store in DB is 

```js
{
    pk: USER#[userID]#INFO
    sk: INFO
    email: [email] 
    passwordSalt: [passwordSalt]
    passwordHash: [passwordHash]
    userName: [userName]
}
```

### CODE REVIEW

File: backend/.env.example [hunk 1]
```diff
@@ -1,2 +1,7 @@
 HOST=localhost
-PORT=3000
\ No newline at end of file
+PORT=3000
+
+# DynamoDB
+AWS_ACCESS_KEY_ID=
+AWS_SECRET_ACCESS_KEY=
```
Comment (yun-si) [inline backend/.env.example:6]: 因為 s3 有一樣的變數名稱
改成
DYNAMODB_ACCESS_KEY_ID
DYNAMODB_SECRET_ACCESS_KEY
比較好～

File: backend/src/controllers/authController.js [hunk 2]
```diff
@@ -1,15 +1,78 @@
...
+            password = req.body.password;
+            if (!password) {
+                res.status(400).json({ message: 'Password is required' });
+                return;
+            }
+            userName = req.body.userName;
+            if (!userName) {
+                res.status(400).json({ message: 'UserName is required' });
+                return;
+            }
```
Comment (yun-si) [inline backend/src/controllers/authController.js]:
```
const { userID, email, password, userName } = req.body;
try {
    if (!userID || !email || !password || !userName) {
        const missingFields = [
            !userID && 'userID',
            !email && 'email',
            !password && 'password',
            !userName && 'userName',
        ].filter(Boolean);

        res.status(401).json({ message: `${missingFields.join(', ')} are required` });
        return;
    }
...
```
這樣寫比較簡潔～ 丟 gpt 產出來的

image

File: backend/src/controllers/authController.js [hunk 3]
```diff
@@ -1,15 +1,78 @@
...
+            email = req.body.email;
+            if (!email) {
+                res.status(400).json({ message: 'Email is required' });
+                return;
+            }
+            password = req.body.password;
+            if (!password) {
+                res.status(400).json({ message: 'Password is required' });
+                return;
+            }
```
Comment (yun-si) [inline backend/src/controllers/authController.js]: 跟上面一樣～

**Review by yun-si — CHANGES_REQUESTED** [review]

File: backend/.env.example (hunk 1 above)
Comment (amyhsu-o) [reply backend/.env.example:6]: 討論後決定 DynamoDB 和 S3 共用一個 IAM user，所以不改

File: backend/src/controllers/authController.js (hunk 2 above)
Comment (amyhsu-o) [reply backend/src/controllers/authController.js]: 改了。統一用 authController.js 裡的 _reqWithMissingValue() 來處理

這招好厲害 😂

File: backend/src/controllers/authController.js (hunk 3 above)
Comment (amyhsu-o) [reply backend/src/controllers/authController.js]: 改了。統一用 authController.js 裡的 _reqWithMissingValue() 來處理

**Review by yun-si — DISMISSED** [review]

**Review by yun-si — APPROVED** [review]

**Pull request merged**